---
title: Commands and Options
---

## NAME

f5c(1) - Ultra-fast methylation calling and event alignment tool for nanopore sequencing data with optional GPU (CUDA) acceleration

## SYNOPSIS

* indexing:
  ```
  f5c index -d [fast5_folder] [read.fastq|fasta]
  ```
* methylation calling:
  ```
  f5c call-methylation -b [reads.sorted.bam] -g [ref.fa] -r [reads.fastq|fasta] > [meth.tsv]
  f5c meth-freq -i [meth.tsv] > [freq.tsv]
  ```
* event alignment:
  ```
  f5c eventalign -b [reads.sorted.bam] -g [ref.fa] -r [reads.fastq|fasta] > [events.tsv]
  ```


## DESCRIPTION

Given a set of base-called nanopore reads and associated raw signals, f5c call-methylation detects the methylated cytosine at genomic CpG cites and f5c eventalign aligns raw nanopore DNA signals (events) to the base-called read. f5c can optionally utilise CUDA enabled NVIDIA graphics cards for acceleration. f5c is a heavily re-engineered and optimised implementation of the call-methylation and eventalign modules in Nanopolish.

## COMMANDS 

* `index`:               
         Build an index that maps read IDs to the corresponding fast5 files (extended nanopolish index).
* `call-methylation`:    
         Classify nucleotides as methylated or not at genomic CpG sites (optimised nanopolish call-methylation).
* `meth-freq`:           
         Calculate methylation frequency at genomic CpG sites (optimised nanopolish calculate_methylation_frequency.py).
* `freq-merge`:          
         Merge multiple methylation frequency tsv files.        
* `eventalign`:          
         Align nanopore events to reference k-mers (optimised nanopolish eventalign).


## OPTIONS

### index

`f5c index [OPTIONS] -d nanopore_raw_file_directory reads.fastq`

Build an index that maps read IDs in a fastq/fasta file to the corresponding fast5 files. f5c index is an extended version of nanopolish index by Jared Simpson. The output of f5c index is equivalent to that from nanopolish index.

*  `-h`, `--help`:                           
   Print the help to the standard out.
*  `-d`, `--directory`:                      
   Path to the directory containing fast5 files. This option can be given multiple times. Both multi-fast5 and single-fast5 are supported. The specified directory will be recursively searched for files ending with the .fast5 extension.
*  `-s`, `--sequencing-summary`:             
   The sequencing summary file generated by Guppy base-caller. This option is not encouraged as inconsistencies in the format of sequencing summary files lead to issues during subsequent steps.
*  `-f`, `--summary-fofn`:                   
   File containing the paths to the sequencing summary files (one per line). This option is not encouraged as inconsistencies in the format of sequencing summary files lead to issues during subsequent steps.
*  `-t INT`:
    Number of threads used for bgzf compression [default value: 1]. Increasing the number of threads makes indexing faster. Ideally, this should be the number of CPU cores.
*  `--iop INT`:
    Number of I/O processes to read fast5 files [default value: 1]. Increasing the number of I/O processes makes indexing significantly faster, especially on HPC with RAID systems (multiple disks) where this can be as high as 64. Note that unless value of iop is 1, options -s and -f  are ignored.
*  `--verbose INT`:
    Verbosity level for the log messages [default value: 0].
*  `--version`:
    Print the version number to the standard out.        


### call-methylation

`f5c call-methylation [OPTIONS] -r reads.fa -b alignments.bam -g genome.fa`

Classify nucleotides as methylated or not at genomic CpG cites (optimised nanopolish call-methylation). Note that the list below contains the options for both CPU-only and CPU-GPU versions of f5c.  Options related to the GPU (CUDA) do NOT apply to the CPU-only version.

#### basic options

* `-r FILE`:                     
  The file containing the base-called reads in FASTQ or FASTA format. Can be gzip compressed files.
* `-b FILE`:                    
  The file contaning the alignment records sorted based on genomic coordinates in BAM format.
* `-g FILE`:                    
  The file containing the reference genome in FASTA format.
* `-w STR`:      
  Only process the specified genomic region STR. STR should be in the format *chr:start-end*. Currently, multiple region strings are not supported. If this option is not specified, the whole genome will be processed.
* `-t INT`:                     
  Number of processing threads [default value: 8]. Ideally, this should be the number of CPU cores.
* `-K INT`:                     
  Maximum number of reads loaded at once to the memory [default value: 512]. A larger value maximises multithreading performance at cost of increased peak RAM.
* `-B FLOAT[K/M/G]`:            
  Maximum number of bases loaded at once to the memory [default value: 2.0M]. A larger value maximises multithreading performance at cost of increased peak RAM.
* `-h`:                         
  Print the help to the standard out.
* `-o FILE`:                    
  The file to write the output. If this option is not specified, the output will be written to the standard out. 
* `-x STR`:                     
  Parameter profile to be used for maximising the performance to a particular computer system. The profile parameters are always applied before other options, i.e., the user can override these parameters explicitly. Some example profiles are laptop, desktop, hpc. See [profiles](https://f5c.page.link/profiles) for the full list and details.
* `--iop INT`:                  
  Number of I/O processes to read FAST5 files [default value: 1]. Increase this value if reading FAST5 limits the overall performance. A higher value (can be as high as 64) is always preferred for systems with multiple disks (RAID) and network file systems.
* `--min-mapq INT`:             
  Minimum mapping quality of an alignment (MAPQ in the BAM record) to be considered for methylation calling [default value: 20].
* `--secondary=yes|no`:         
  Whether secondary alignments are considered or not for methylation calling [default value: no].
* `--verbose INT`:              
  Verbosity level for the log messages [default value: 0].
* `--version`:                  
  Print the version number to the standard out.
* `--disable-cuda=yes|no`:      
  Disable running on the GPU or not [default value: no]. If this option is set to yes, GPU acceleration is disabled. 
* `--cuda-dev-id INT`:          
  CUDA device identifier to run GPU kernels on [default value: 0]. The device identifier of the first GPU is 0, the second GPU is 1 and so on. This can be found by invoking the  `nvidia-smi` command. Currently, only a single GPU can be specified. To utilise multiple GPUs, you have to manually invoke multiple f5c commands on different datasets with a different device identifier.  
* `--cuda-max-lf FLOAT`:        
  Process reads with read-length less than or equal to the product of *cuda-max-lf* and the average read length in the current batch on GPU. The rest is processed on CPU [default value: 3.0]. Useful for tuning the CPU-GPU load balance for atypical datasets. Refer to [performance guidelines](https://hasindu2008.github.io/f5c/docs/f5c-perf-hints) for details.
* `--cuda-avg-epk FLOAT`:       
  The average number of events-per-kmer used for allocating the arrays in GPU memory [default value: 2.0]. Useful for tuning the CPU-GPU load balance for atypical datasets. Refer to [performance guidelines](https://hasindu2008.github.io/f5c/docs/f5c-perf-hints) for details.
* `--cuda-max-epk FLOAT`:       
  Process the reads with events-per-kmer less than or equal to *cuda_max_epk* on GPU. The rest is processed on CPU [default value: 5.0]. Useful for tuning the CPU-GPU load balance for atypical datasets. Refer to [performance guidelines](https://hasindu2008.github.io/f5c/docs/f5c-perf-hints) for details.

#### advanced options

* `--skip-ultra FILE`:    
  Skip ultra-long reads and write those alignment entries to the bam file provided as the argument. Ultra-long reads refer to reads longer than 100 kbases by default, unless specified by --ultra-thresh option below. Useful for tuning the CPU-GPU load balance for datasets containing many ultra-long reads. Also useful to cap the peak RAM usage in systems with limited memory. After the execution, ultra-long reads cab be separately processed, i.e., f5c can be again invoked on the produced bam file as the input.  Refer to [performance guidelines](https://hasindu2008.github.io/f5c/docs/f5c-perf-hints) for details. 
* `--ultra-thresh INT`:  
  Threshold to skip ultra-long reads [default value: 100000]. This option is to be used in conjunction with  `--skip-ultra` above.
* `--skip-unreadable=yes|no`:  
  Whether to skip any unreadable fast5 files or to terminate the program [default value: yes]. If `yes`, the programme will continue to run while skipping unreadable fast5 files. If `no`, the programme will terminate with an error when an unreadable fast5 file is found.
* `--kmer-model FILE`:  
  Custom nucleotide 6-mer model file. The file should adhere to the format in [r9.4_450bps.nucleotide.6mer.template.model](https://github.com/hasindu2008/f5c/blob/master/test/r9-models/r9.4_450bps.nucleotide.6mer.template.model). 
* `--meth-model FILE`:  
  custom methylation 6-mer model file. The file should adhere to the format in [r9.4_450bps.cpg.6mer.template.model](https://github.com/hasindu2008/f5c/blob/master/test/r9-models/r9.4_450bps.cpg.6mer.template.model).
* `--meth-out-version INT`:  
  Format version of the output Methylation tsv file. If set to 1, the columns printed adhere to the output format of Nanopolish early versions. If set to 2, adhere to the latest nanopolish output format that additionally includes the strand column and the header num_cpgs renamed to *num_motifs*) [default value: 1]
* `--cuda-mem-frac FLOAT`:  
  Fraction of free GPU memory to allocate [default value: 0.9 for non-tegra GPUs and 0.7 for tegra GPUs]. On GPUs with dedicated RAM (e.g., GeForce, Tesla and Quadro) almost all available free GPU memory can be allocated. A slightly lower value such as 0.9 is preferred instead of 1.0 to prevent unexpected crashes. In GPUs with integrated memory shared with RAM (e.g., Tegra GPUs that are in Jetson boards), this value should be at most 0.7 to allow enough free RAM for both f5c and other programmes.


#### developer options

* `--print-events=yes|no`:  
  Print the event table (the output of the event detection step) to the standard out.
* `--print-banded-aln=yes|no`:  
  Print the event alignment (the output of the adaptive banded event alignment step) to the standard out.
* `--print-scaling=yes|no`:  
  Prints the estimated scaling values to the standard out.
* `--print-raw=yes|no`:  
  Prints the raw signal to the standard out.
* `--debug-break INT`:  
  Terminate the programme after processing the specified batch number. E.g., If 0 is specified, the programme breaks after processing the 0th batch.
* `--profile-cpu=yes|no`:  
  Process section by section and separately print the time spent on different steps such as the event detection, ABEA and HMM. This option is used for profiling the workloads on the CPU. 
* `--write-dump=yes|no`:  
  Write the fast5 dump to a file or not. The file name is hardcoded to f5c.tmp.bin and will be written to the current working directory. The required raw signal data in the fast5 files subsequent processing will be serially written to *f5c.tmp.bin*.
* `--read-dump=yes|no`:  
  Read from a fast5 dump file or not. This is used to read from a dump file generated using `--write-dump` above. The raw signal data will be serially loaded from the dump file instead of the fast5 files.
         
### meth-freq

Calculate methylation frequency at genomic CpG sites from a tsv file containing per-read methylation calls (optimised nanopolish calculate_methylation_frequency.py). Both tsv formats (with/without strand column and/or num_cpg/num_motif) are supported and is detected automaticaly.

`meth-freq [options...]`

```
  -c float]         Call threshold. Default is 2.5.
  -i [file]         Input tsv file containing per-read methylation calls. Read from stdin if not specified.
  -o [file]         Output file. Write to stdout if not specified.
  -s                Split groups
  ```

### freq-merge

 merge multiple methylation frequency tsv files to a single tsv file. Useful to combine the results when meth-freq was run separately on batches, for instance, when performing methylation calling on-the-fly.
For each methylation calling output (.tsv) file, perform meth-freq separately (no concatenation required). Then feed those output (.tsv) files to this tool, to obtain the final methylation frequency calculated file. 

```
freq-merge [options...]
  -o FILE           Output file.
  -n [INT]          Number of methylation frequency .tsv files to be merged
  -f                n number of input filepaths should be followed
  e.g. freq-merge -o merged_freq.tsv -n 2 -f data1_freq.tsv data2_freq.tsv 
  ```
=======

### eventalign

`f5c eventalign [OPTIONS] -r reads.fa -b alignments.bam -g genome.fa`

```
basic options:
   -r FILE                    fastq/fasta read file
   -b FILE                    sorted bam file
   -g FILE                    reference genome
   -w STR                     limit processing to genomic region string of format chr:start-end
   -t INT                     number of processing threads [8]
   -K INT                     batch size (max number of reads loaded at once) [512]
   -B FLOAT[K/M/G]            max number of bases loaded at once [5.0M]
   -h                         help
   -o FILE                    output to file [stdout]
   -x STR                     parameter profile to be used for better performance (always applied before other options)
                              e.g., laptop, desktop, hpc; see https://f5c.page.link/profiles for the full list
   --iop INT                  number of I/O processes to read fast5 files [1]
   --min-mapq INT             minimum mapping quality [20]
   --secondary=yes|no         consider secondary mappings or not [no]
   --verbose INT              verbosity level [0]
   --version                  print version

advanced options:
   --skip-ultra FILE          skip ultra long reads and write those entries to the bam file provided as the argument
   --ultra-thresh INT         threshold to skip ultra long reads [100000]
   --skip-unreadable=yes|no   skip any unreadable fast5 or terminate program [yes]
   --kmer-model FILE          custom nucleotide 6-mer model file (format similar to test/r9-models/r9.4_450bps.nucleotide.6mer.template.model)
   --summary FILE             summarise the alignment of each read/strand in FILE
   --sam                      write output in SAM format
   --print-read-names         print read names instead of indexes
   --scale-events             scale events to the model, rather than vice-versa
   --samples                  write the raw samples for the event to the tsv output

developer options:
   --print-events=yes|no      prints the event table
   --print-banded-aln=yes|no  prints the event alignment
   --print-scaling=yes|no     prints the estimated scalings
   --print-raw=yes|no         prints the raw signal
   --debug-break INT          break after processing the specified no. of batches
   --profile-cpu=yes|no       process section by section (used for profiling on CPU)
   --write-dump=yes|no        write the fast5 dump to a file or not
   --read-dump=yes|no         read from a fast5 dump file or not
```   

## EXAMPLES


* download and extract the dataset including sorted alignments:
  ```
  wget -O f5c_na12878_test.tgz "https://f5c.page.link/f5c_na12878_test"
  tar xf f5c_na12878_test.tgz
  ```
* index, call methylation and get methylation frequencies:
   ```
   f5c index -d chr22_meth_example/fast5_files chr22_meth_example/reads.fastq
   f5c call-methylation -b chr22_meth_example/reads.sorted.bam -g chr22_meth_example/humangenome.fa -r chr22_meth_example/reads.fastq > chr22_meth_example/result.tsv
   f5c meth-freq -i chr22_meth_example/result.tsv > chr22_meth_example/freq.tsv
   ```
* event alignment:
  ```
  f5c eventalign -b chr22_meth_example/reads.sorted.bam -g chr22_meth_example/humangenome.fa -r chr22_meth_example/reads.fastq > chr22_meth_example/events.tsv
  ```

## AUTHOR

Hasindu Gamaarachchi wrote the framework of f5c, CUDA code and integrated with adapted components from Jared T. Simpson's Nanopolish [https://github.com/jts/nanopolish], with tremendous support from Chun Wai Lam, Gihan Jayatilaka and Hiruna Samarakoon. 

## LICENSE

f5c is licensed under the MIT License. f5c reuses code and methods from [Nanopolish](https://github.com/jts/nanopolish) which is also under the MIT License. The event detection code in f5c is from Oxford Nanopore's [Scrappie basecaller](https://github.com/nanoporetech/scrappie) which is under Mozilla Public License 2.0. Some code snippets have been taken from [Minimap2](https://github.com/lh3/minimap2) and [Samtools](https://github.com/samtools/samtools) that are under the MIT License.

If you use f5c, please cite Gamaarachchi, H., Lam, C.W., Jayatilaka, G. et al. GPU accelerated adaptive banded event alignment for rapid comparative nanopore signal analysis. BMC Bioinformatics 21, 343 (2020). https://doi.org/10.1186/s12859-020-03697-x

## SEE ALSO

Full documentation: https://hasindu2008.github.io/f5c/docs/overview

Source code: https://github.com/hasindu2008/f5c/

Publication: https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-020-03697-x

